segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkStorageManager.java

```diff
@@ -0,0 +1,1350 @@
+                val readIndex = getReadIndex(streamSegmentName);
+                if (readIndex.size() > 0) {
+                    val floorEntry = readIndex.floorEntry(offset);
+                    if (null != floorEntry) {
+                        startOffsetForCurrentChunk = floorEntry.getKey();
+                        currentChunkName = floorEntry.getValue();
+                    }
+                }
+
+                // Navigate to the chunk that contains the first byte of requested data.
+                while (currentChunkName != null) {
+                    chunkToReadFrom = (ChunkMetadata) txn.get(currentChunkName);
+                    Preconditions.checkState(null != chunkToReadFrom, "chunkToReadFrom is null");
+                    if (   startOffsetForCurrentChunk <= currentOffset
+                        && startOffsetForCurrentChunk + chunkToReadFrom.getLength() > currentOffset) {
+                        // we have found a chunk that contains first byte we want to read
+                        log.debug("read: found chunk to read segment: {} chunk: {}", streamSegmentName, chunkToReadFrom);
+                        break;
+                    }
+                    currentChunkName = chunkToReadFrom.getNextChunk();
+                    startOffsetForCurrentChunk += chunkToReadFrom.getLength();
+
+                    // Update read index with newly visited chunk.
+                    if (null != currentChunkName) {
+                        readIndex.put(startOffsetForCurrentChunk, currentChunkName);
+                    }
+                }
+
+                // Now read.
+                while (bytesRemaining > 0 && null != currentChunkName) {
+                    int bytesToRead = Math.min(bytesRemaining, Math.toIntExact(chunkToReadFrom.getLength() - (currentOffset - startOffsetForCurrentChunk)));
+                    //assert bytesToRead != 0;
+
+                    if (currentOffset >= startOffsetForCurrentChunk + chunkToReadFrom.getLength()) {
+                        // The current chunk is over. Move to the next one.
+                        currentChunkName = chunkToReadFrom.getNextChunk();
+                        if (null != currentChunkName) {
+                            startOffsetForCurrentChunk += chunkToReadFrom.getLength();
+                            chunkToReadFrom = (ChunkMetadata) txn.get(currentChunkName);
+                            log.debug("read: reading from next chunk segment: {} chunk: {}", streamSegmentName, chunkToReadFrom);
+                        }
+                    } else {
+                        Preconditions.checkState(bytesToRead != 0, "bytesToRead is 0");
+                        // Read data from the chunk.
+                        ChunkHandle chunkHandle = chunkStorage.openRead(chunkToReadFrom.getName());
+                        int bytesRead = chunkStorage.read(chunkHandle, currentOffset - startOffsetForCurrentChunk, bytesToRead, buffer, currentBufferOffset);
```

Comment:
What about supporting a chunkStorage that is able to perform async read operations ?

This method looks like it is capable of working with an async data source as it returns a CompletableFuture<Integer> but is actually performing blocking reads.

Reply:
That was the original plan. However, the required changes were getting bit complicated.  I wasn't sure this added complexity was actually buying a lot of performance. (Java threads doing IO do get blocked anyway. ) So I decided to do it in small steps. In this iteration still keeping the sync interface. In next iteration or two I do plan to turn this into async operation.